spec/thread_safety_spec.rb

```diff
@@ -16,20 +14,65 @@
       Class.new do
         prepend MemoWise
 
-        def first_thread_id
-          Thread.pass # trigger a race condition even on MRI
-          Thread.current.object_id
+        def current_thread_id
+          Thread.pass              # trigger a race condition even on MRI
+          Thread.current.object_id # return different values in each thread
         end
-        memo_wise :first_thread_id
+        memo_wise :current_thread_id
       end
     end
 
     # Tip of the hat to @jeremyevans for finding and fixing a race condition
     # which caused accidental `nil` values to be returned under contended calls
     # to unmemoized zero-args methods.
     #   * See https://github.com/panorama-ed/memo_wise/pull/224
-    it "does not return accidental nil value to either thread" do
-      expect(thread_return_values).not_to include(nil)
+    context "when checking condition: are nil values accidentally returned?" do
+      let(:condition_to_check) do
+        ->(values) { values.any?(&:nil?) }
+      end
+
+      it "does not return accidental nil value to either thread" do
+        expect(thread_return_values).not_to include(nil)
+      end
+    end
+
+    # When memoizing non-deterministic methods, MemoWise does not impose any
+    # additional guarantees other than:
+    #   * Return values are one of the valid returns from the original method
+    #   * Return values differ between threads under contended calls
+    #   * One of those return values will be memoized and returned thereafter
+    context "when checking condition: are different values returned?" do
+      let(:condition_to_check) do
+        ->(values) { values.any? { |v| v != values.first } }
+      end
+
+      let(:after_memoization_thread_return_values) do
+        _ = thread_return_values # sanity check - always comes after first `let`
+
+        check_repeatedly(condition_proc: condition_to_check) do
+          threads = Array.new(2) { Thread.new { @instance.current_thread_id } } # rubocop:disable RSpec/InstanceVariable
+          threads.map(&:value)
+        end
+      end
+
+      # NOTE: Disabled on TruffleRuby in CI, because it takes multiple seconds
+      # to observe the different values returned to multiple threads, and errors
+      # out in some cases when that does happen.
+      unless RUBY_ENGINE == "truffleruby" && ENV["CI"] == "true"
+        it "returns different values to each thread, and memoizes one of them" do
+          # Before memoization: expect to observe threads return different values
+          expect(thread_return_values).
```

Comment:
Similar to my comment above, would this be clearer?

```ruby
expect(thread_return_values.uniq.size).to be > 1
```

I found the `include(not_eq(... .first))` difficult to understand.

Reply:
Changed! Thanks